qhbmlib/inference/qnn.py

```diff
@@ -121,22 +121,95 @@ def _sampled_expectation(self, initial_states: tf.Tensor,
       unaveraged expectation values of `observable` against each transformed
       initial state.
     """
-    u = self.circuit + observable.circuit_dagger
-    unique_states, idx, _ = utils.unique_bitstrings_with_counts(initial_states)
-    circuits = u(unique_states)
-    num_circuits = tf.shape(circuits)[0]
-    tiled_values = tf.tile(
-        tf.expand_dims(u.symbol_values, 0), [num_circuits, 1])
-    samples = self._sample_layer(
-        circuits,
-        symbol_names=u.symbol_names,
-        symbol_values=tiled_values,
-        repetitions=self._expectation_samples).to_tensor()
-    unique_expectations = tf.map_fn(
-        lambda x: tf.math.reduce_mean(observable.energy(x)),
-        samples,
-        fn_output_signature=tf.float32)
-    return utils.expand_unique_results(unique_expectations, idx)
+
+    @tf.custom_gradient
+    def _inner_expectation():
+      """Enables derivatives."""
+      unique_states, idx, _ = utils.unique_bitstrings_with_counts(
+          initial_states)
+      u = self.circuit + observable.circuit_dagger
```

Comment:
nit: this is not necessarily unitary so wouldn't call it `u`

Reply:
We do assume it is unitary, otherwise the gradient result is not correct.  But I agree `u` is not an informative name, changed to `total_circuit`

Reply:
The observable can't be a general Hermitian operator?

Reply:
We assume the operator is diagonalized into a unitary circuit (`observable.circuit`) and a diagonal (`observable.energy`).  Every Hermitian operator can be represented this way, so it is fully general

Reply:
Thanks for clarifying - it means that the observable is represented as QHBM. I missed this aspect again where it seemed like a class could be any general representation and not necessarily spectral

Reply:
My suggestion would be to at least update the docstring with the parameterization specified